spock-core/src/main/java/org/spockframework/runtime/SpockEngineDiscoveryPostProcessor.java

```diff
@@ -14,6 +16,7 @@ SpockEngineDescriptor postProcessEngineDescriptor(UniqueId uniqueId, RunContext
     SpockEngineDescriptor processedEngineDescriptor = new SpockEngineDescriptor(uniqueId, runContext);
     engineDescriptor.getChildren().stream()
       .map(child -> processSpecNode(child, runContext))
+      .sorted(comparingInt(child -> child instanceof SpecNode ? ((SpecNode) child).getNodeInfo().getExecutionOrder() : 0))
```

Comment:
The order will not change. 0 is also the default value for the property in `SpecInfo`.

Reply:
Try commenting out https://github.com/spockframework/spock/blob/8add40440647bc974dc25e339050e415cdb405ba/spock-specs/src/test/groovy/org/spockframework/smoke/extension/RandomRunOrderExtensionSpec.groovy#L67 in the deterministic order test. it will still pass.

Reply:
> Furthermore, it may work for randomness but how would you do an explicit execution order of Specs?

I think this is the bigger issue, with the way it is implemented for this PR you can't really do anything other than randomize the results, because you don't get access to all the `SpecInfo` instances to compare them against each other, as opposed to the existing features.

Reply:
I think you misunderstood me, what I mean is that the current `visitSpec` is not really well suited to set the execution orders across specs. For example you can't directly compare two `SpecInfo`, making it impossible to implement alphabetical sorting in an efficient way. I think a better alternative would be to make it possible to define a single `Comparator<SpecInfo>` for a runner. This way we can make the sorting step optional as well, i.e. only do it if the comparator is actually set.

Reply:
Ah yes, thanks for the feedback. That is assuming that we actually want to enable users to implement something like alphabetical order, which I see zero value in, or at least not enough value to justify changing the inner workings of Spock for it. My change is minimally invasive and would cater to both randomised and manual, annotation-driven ordering.

Besides, the comparator would not be well suited for random ordering, if you really want to implement a good comparator obeying transitivity, i.e. `a < b` and `b < c` also means `a < c`. The simple random ordering use case, i.e. the most valuable one for real world testing, would require the most infrastructure to implement.

My recommendation is to simply disregard and not support that use case. If anyone wants it, it can still be implemented in an inefficient way. 😉 Think collecting all `SpecInfo`s in a static field of the global extension during `visitSpec`, repeatedly updating all orders. Slow and ugly for very many specs, but it works:

```groovy
import org.spockframework.runtime.extension.IGlobalExtension
import org.spockframework.runtime.model.FeatureInfo
import org.spockframework.runtime.model.SpecInfo

class AlphabeticalOrderExtension implements IGlobalExtension {
  private static SortedSet<SpecInfo> specInfos = 
    new TreeSet<>({ SpecInfo left, SpecInfo right -> right.displayName.compareTo(left.displayName) })

  @Override
  void visitSpec(SpecInfo spec) {
    specInfos << spec
    int i = 0
    for (SpecInfo specInfo : specInfos)
      specInfo.executionOrder = i++
    spec.features
      .sort { FeatureInfo left, FeatureInfo right -> right.displayName.compareTo(left.displayName) }
      .eachWithIndex { FeatureInfo entry, int order -> entry.executionOrder = order }
  }
}
```

Of course, it would be easier if there was an extra event `afterVisitSpecs` for global extensions after visiting the last spec. Then, we could collect the specs in `visitSpec` and assign the order just once in `afterVisitSpecs` and also expunge the sorted set after doing so, getting rid of the memory leak. But like I said, this use case for me has no value whatsoever, so I would deliberately decide not to support it at all.

If of course you want to support all thinkable future use cases, then we need all the extra infrastructure you are thinking about. But, comparing cost and benefit, I suggest to just drop it.

Reply:
> the current `visitSpec` is not really well suited to set the execution orders across specs.

I think that the new lifecycle method `IGlobalExtension.initSpecs` addresses this issue.